hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/HandleWorkflow.java

```diff
@@ -476,6 +488,51 @@ private void handleUserTransaction(
         blockRecordManager.endUserTransaction(recordListResult.records().stream(), state);
     }
 
+    /**
+     * Updates key on the hollow accounts that need to be finalized. This is done by dispatching a preceding
+     * synthetic update transaction. The ksy is derived from the signature expansion, by looking up the ECDSA key
+     * for the alias.
+     *
+     * @param context the handle context
+     * @param configuration the configuration
+     * @param accounts the set of hollow accounts that need to be finalized
+     * @param verifier the key verifier
+     */
+    private void finalizeHollowAccounts(
+            @NonNull final HandleContext context,
+            @NonNull final Configuration configuration,
+            @NonNull final Set<Account> accounts,
+            @NonNull final DefaultKeyVerifier verifier) {
+        final var consensusConfig = configuration.getConfigData(ConsensusConfig.class);
+        final var precedingHollowAccountRecords = accounts.size();
+        final var maxRecords = consensusConfig.handleMaxPrecedingRecords();
+        // If the hollow accounts that need to be finalized is greater than the max preceding
+        // records allowed throw an exception
+        if (precedingHollowAccountRecords >= maxRecords) {
+            throw new HandleException(MAX_CHILD_RECORDS_EXCEEDED);
+        } else {
+            for (final var hollowAccount : accounts) {
+                // get the verified key for this hollow account
+                final var verification = verifier.verificationFor(hollowAccount.alias());
+                if (verification.key() != null) {
+                    if (!IMMUTABILITY_SENTINEL_KEY.equals(hollowAccount.keyOrThrow())) {
+                        logger.error("Hollow account {} has a key other than the sentinel key", hollowAccount);
+                        return;
```

Comment:
Should we throw a HandleException instead of just return? We would have to check the log to find out what went wrong.

Reply:
yes makes sense. Not sure on the response code 🤔

Reply:
Maybe KEY_REQUIRED?

Reply:
DevOps has alerts on all `ERROR` logs so we would be notified of this.

But imo we would throw an `IllegalStateException` here because this should be impossible ([`isHollow()`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/HapiUtils.java#L92) only classifies an account as "hollow" if its key is the `IMMUTABILITY_SENTINEL_KEY`).